issue-1132 - refactor multi tenancy and schema migration docs (#1165)
IBM/FHIR: fhir-persistence-schema/docs/DB2MultiTenancy.md

# The IBM FHIR Server - Multi-Tenancy on Db2

> This document outlines the schema design and implementation for the IBM FHIR Server's Multi-tenancy feature on Db2.

Reviewer: Where is tenant_query_rewrite.png used?

Author: Not referenced at this time, but will be useful for helping explain the Db2 row permission feature more clearly.

Reviewer: I'm not sure how I feel about something unreferenced. Maybe just add it inline>?

The Multi-tenancy feature has two main areas of focus: 
- Tenant Provisioning  
- Security

This document assumes familiarity with the general [IBM FHIR Server schema design](SchemaMigration.md).

# Multi-Tenancy Concepts and Design

The IBM FHIR Server design supports three levels of multi-tenancy:

1. Separate database per tenant
2. Separate data schema per tenant

> 3. (Db2-only) multiple tenants per schema, partitioned with row-based access control and per-tenant tablespaces.

Author: I'll make this more clear. We do support 1 and 2 because each tenant gets its own fhir-server-config datasource set up. We can even mix database types.

> 3. (Db2-only) multiple tenants per schema, partitioned with row-based access control and per-tenant tablespaces.

Author: done

Of course, an entire deployment (application servers and database) also could be dedicated to a single tenant, which is really just a variant of option 1 above, although clearly such a configuration would not be considered multi-tenant.

## Tenant Data Administration

> The IBM FHIR Server separates schema administration data from user data. All administration data is held in tables within the FHIRADMIN schema, and these tables are not directly accessible by the IBM FHIR Server database user.

Author: fixed

> The FHIRADMIN schema is used to manage schema version migrations as well as tenant configuration for any tenants resident in the given database. The FHIRADMIN schema does not manage tenants in other databases.

Author: fixed

See the general [IBM FHIR Server schema design](SchemaMigration.md) documentation for a more detailed description of how FHIRADMIN tables are used to manage schema versions.

## Multi-tenancy using Db2 Table Partitions and Per Tenant Tablespaces

The IBM FHIR Server provides a multi-tenancy feature based on Db2 row-based access control and table partitioning features.

Each new tenant is allocated a unique tenant id value which is of database type `INT`. All tables in the IBM FHIR Server data schema (typically `FHIRDATA`) are created as range partitioned and configured with row-based access control. 

As an example, consider a table of patients created using the following DDL (illustrative only, this is not the actual IBM FHIR Server schema):

```

> CREATE TABLE ptng.patients (

Author: This came from my proof of concept and I wanted to avoid specifying DDL which someone might run and create objects in our real FHIR schema.

    mt_id               INT     NOT NULL,
    patient_name    VARCHAR(18) NOT NULL,
    age                 INT     NOT NULL,
    doctor_name     VARCHAR(18),
    ssn                CHAR(11),
    CONSTRAINT pk_patient PRIMARY KEY (mt_id, patient_name)
)
PARTITION BY RANGE (mt_id) (STARTING 0 INCLUSIVE ENDING 0 INCLUSIVE);
```

Each tenant is allocated a unique integer value for mt_id (multi-tenant id), with a range partition used to define which table partition is to be used for a particular tenant's mt_id. The mapping of tenant name to mt_id value is maintained in the FHIR_ADMIN.TENANTS table.

The above example assumes that patient_name will be unique within the context of a single tenant, which obviously does not apply to the actual IBM FHIR Server schema. The primary key must include the partition range key `mt_id`. In addition, any secondary indexes MUST also include mt_id to ensure they are always local to the partition. For example:

```

> CREATE UNIQUE INDEX ptng.unq_ssn ON ptng.patients(mt_id, ssn);

Reviewer: ```suggestion
CREATE UNIQUE INDEX fhirdata.unq_ssn ON fhirdata.patients(mt_id, ssn);
```

Author: ended up changing the schema name to "example" to make it clearer.

Each tenant is allocated a partition based on the MT_ID and assigned to a tablespace. The tenant `tablespace` is created one time and no changes to the tablespace are expected.

In general, smaller extent sizes can harm the performance of table and index scans because they limit the amount of data which can be read in a single IO operation. However, such scans are not common in most FHIR query use-cases, so this tends to not be an issue.

## Multi-tenant Queries

We could use the multi-tenant enabled table definition as-is and modify every SQL statement to include `AND mt_id = ?` but this is error-prone and risky. If a query was missed or poorly written and ran without that predicate, it could expose data from another tenant.

Instead, we leverage DB2 permission objects which allow us to specify a filter condition. Each query submitted by the application is rewritten by the database to include the `FOR ROWS WHERE` predicate, which is defined as: 

```
CREATE PERMISSION ROW_ACCESS ON ptng.patients
         FOR ROWS WHERE patients.mt_id = ptng.session_tenant
         ENFORCED FOR ALL ACCESS
           ENABLE;

> ALTER TABLE ptng.patients ACTIVATE ROW ACCESS CONTROL;

Author: schema name is now example

```

The `session_tenant` above is defined as a session variable. The application must set the value to match the mt_id value configured for the tenant.

Logically:

```

> SET ptng.session_tenant = 1;
> SELECT * FROM ptng.patients;
> commit;
> ```
> 
> Internally, DB2 rewrites the query to look like this:
> 
> ```
> SELECT * FROM ptng.patients WHERE mt_id = 1;
> ```
> 
> Because the table `ptng.patients` is partitioned using mt_id, the query optimizer is able to perform a partition-prune, by limiting the query access to the partition associated with the tenant. This results in a query execution plan very similar in terms of performance to a single-tenant scenario without any partitioning.

Reviewer: ```suggestion
SET fhirdata.session_tenant = 1;
SELECT * FROM fhirdata.patients;
commit;
```

Internally, DB2 rewrites the query to look like this:

```
SELECT * FROM fhirdata.patients WHERE mt_id = 1;
```

Because the table `fhirdata.patients` is partitioned using mt_id, the query optimizer is able to perform a partition-prune, by limiting the query access to the partition associated with the tenant. This results in a query execution plan very similar in terms of performance to a single-tenant scenario without any partitioning.
```

Author: schema name is now "example"

To isolate a tenant's data at the file-system level, a tablespace is created per tenant and the partition objects are are allocated to that tablespace, including any local indexes:

```
CREATE TABLESPACE TS_TENANT1 MANAGED BY AUTOMATIC STORAGE;
CREATE TABLESPACE TS_TENANT2 MANAGED BY AUTOMATIC STORAGE;
```

Adding new tenants requires adding partitions to each table in the data schema. Note how the partition is configured with the tenant's specific tablespace:

```
-- for tenant #1
ALTER TABLE ptng.patients ADD PARTITION TENANT1 STARTING 1 INCLUSIVE ENDING 1 INCLUSIVE IN TS_TENANT1;

-- for tenant #2

> ALTER TABLE ptng.patients ADD PARTITION TENANT2 STARTING 2 INCLUSIVE ENDING 2 INCLUSIVE IN TS_TENANT2;

Author: schema name is now "example"

The process to delete a tenant is (logically):

```
ALTER TABLE <tbl> DETACH PARTITION INTO <tbl_drp1>;
...
COMMIT;
DROP TABLE <tbl_drp1>;
...
COMMIT;
DROP TABLESPACE TS_TENANT1;
```

In this case, TS_TENANT1 is the name of the tablespace previously assigned to the tenant being dropped.

Note that DB2 requires a commit after the partition is detached and before the table `<tbl_drp1>` can be dropped. The tooling must implement this process to be idempotent to handle cases where a detach or drop might fail.

Dropping partitions in this way works by swizzling metadata in the catalog which requires very little (redo) logging and is therefore far more efficient than deleting actual rows of data.

> 

Author: fixed

## Tenant Authentication

The server connect to Db2 with a user configured for the least privileges it requires to provide the FHIR resource persistence and search functions. This user does not have any database administration privileges, unlike the FHIRADMIN user which is used to deploy the schema, allocate tenants etc.

> ![Privilege Design](privilege_design.png)

Reviewer: This image needs some work. 

Schema is FHIR_ADMIN
I think the red lines should be removed
the CREATE PERMISSIONS should be removed at the bottom

The following table summarizes the properties/values involved in authenticating tenant access:

| Property | Description |
|----------|-------------|
|TENANT_NAME | Used by applications to uniquely identify tenants |
| MT_ID | The integer primary key of the TENANTS table, used to map the TENANT_NAME to an integer value which is used as the partitioning key and row-based access permission predicate. |
| TENANT_KEY | A 32 byte random value created when the tenant is provisioned. This is a secret, and must be passed by the client to the database in order to obtain access to a tenant. A tenant may have multiple tenant keys to support key rotation. (deploy new key before removing old key) |
| TENANT_SALT | A 32 byte random value used as a salt when computing the SHA-256 tenant hash |
| TENANT_HASH | A SHA-256 hash of the TENANT_KEY and TENANT_SALT. The TENANT_KEY is never stored directly by the database. |

> The FHIR_ADMIN.SET_TENANT stored procedure takes two parameters: the tenant name and the tenant key. This acts as a user/password pair tuple. The SET_TENANT code looks up a previously stored salt value associated with the tenant name and computes a hash of the salt + tenant key. If the result matches the stored hash value, the call is considered authentic, and the session_tenant variable is set.

Reviewer: ```suggestion
The FHIR_ADMIN.SET_TENANT stored procedure takes two parameters: the tenant name and the tenant key. This acts as a authorization. The SET_TENANT code looks up a previously stored salt value associated with the tenant name and computes a hash of the salt + tenant key. If the result matches the stored hash value, the call is considered authentic, and the session_tenant variable is set.
```

I suggest minimizing the user/password reference, a user reading this may conflate it with their user-password.

Author: fixed

To summarize, for a client application to access tenant data in the multi-tenant Db2 schema, the following must apply:

> 1. The client (e.g. the server) must have access to the database as the FHIRUSER user;

Reviewer: ```suggestion
1. The client (e.g. the server) must have access to the database as the FHIRSERVER user;
```

Reviewer: please use FHIRSERVER

Author: all changed

2. The user FHIRUSER must be granted execute privilges on the FHIR_ADMIN.SET_TENANT procedure;
3. The user FHIRUSER must be granted READ on FHIR_ADMIN.SV_TENANT_ID;
4. The user FHIRUSER must be granted SELECT,INSERT,UPDATE,DELETE on all tables in the FHIRDATA schema. 

## Summary of Granted Privileges:

| Grants |
| --- |
| All objects in schema FHIR_ADMIN are owned by user FHIRADMIN |
| All objects in schema FHIRDATA are owned by user FHIRADMIN |
| GRANT EXECUTE ON FHIR_ADMIN.SET_TENANT TO FHIRUSER; |
| GRANT READ ON FHIR_ADMIN.SV_TENANT_ID TO FHIRUSER; |
| GRANT EXECUTE ON FHIRDATA.ADD_RESOURCE_TYPE TO FHIRUSER; |
| GRANT EXECUTE ON FHIRDATA.ADD_RESOURCE TO FHIRUSER; |
| GRANT SELECT,INSERT,UPDATE,DELETE ON FHIRDATA.PATIENT_RESOURCES TO FHIRUSER; |

## Tenant Data Security

> HIPAA controls require protected data at rest is encrypted, and this is implemented by using Db2 native encryption. When the tenant is deleted, its data must be rendered inaccessible. This is achieved by the following:

Reviewer: I don't think we should mention HIPAA.

Author: changed to PHI

1. Detaching the tenant's partition and dropping the table created by the detach operation.

> 2. Dropping the tenant's tablespace, which will delete the database container files on the underlying OS filesystem.

Reviewer: ```suggestion
2. Dropping the tenant's tablespace, which deletes the database container files on the underlying OS filesystem.
```

Author: ok

The above actions mean that the tenant's data would no longer be accessible at the database or file-system levels.

In the past, there may have been an argument for a third step involving scrubbing the files by overwriting their data with zeroes or random values. However, in today's data centers with heavily virtualized storage sub-systems, there would be no guarantee that such a write would overwrite the actual on-disk data blocks anyway (storage sub-systems may use copy-on-write techniques or similar approaches to provide snapshots, for example). In addition, when consuming Db2 as a cloud service, OS-level access is unlikely to be available, making a file scrub impossible, unless it was exposed by the cloud provider as a feature of the service.

Because dropping a tablespace causes Db2 to delete the underlying container files, any attempt to recover the data would require:

1. privileged access to the OS or storage sub-system;
2. the database encryption key;
3. detailed knowledge of OS file systems and data recovery tools;
4. detailed knowledge of Db2 internals and data recovery tools;
5. opportunity to retrieve the data before the OS reused the free blocks in the filesystem.

> 
> 
> 

Reviewer: ```suggestion
```

Author: ok

The tables have supporting indicies and privileges. 

**Table: VERSION_HISTORY**

Before any table or schema object is created, the `VERSION_HISTORY` table is created using `CreateVersionHistory.createTableIfNeeded(adminSchemaName, adapter);`  This table includes VERSION_HISTORY for resources created in the schema. If an object in the schema is updated, then the VERSION_HISTORY table must be updated to track the changes. 

The `VERSION_HISTORY` table enables Tenant version isolation, so upgrades and patches are separately applied to the schema to which the tenant belongs. Specifically, if the tenant shares the table definitions with another client, the patching is applied to both tenants at the same time.  If the tenant has a specific schema, the table definitions may be updated independent of each other.

The `CreateVersionHistory` class controls the creation the `VERSION_HISTORY` table. Importantly, the VERISON_HISTORY table does not support migrations and changes. The Primary Key has a corresponding index `PK_VERSION_HISTORY`, and the code does not support updating the index. 

This table is an administrative table and should not require updating and migration. 

> 
> 
> 

Reviewer: ```suggestion
```

Author: fixed

| VERSION | INTEGER | 4 | No | The version starting at one and incrementing with every version change |
| APPLIED | TIMESTAMP | 10 | No | The time of the latest object update |

> ## SV_TENANT_ID SESSION VARIABLE 

Reviewer: ```suggestion
**SESSION VARIABLE: SV_TENANT_ID**
```

Author: fixed

The `FHIR_ADMIN.SV_TENANT_ID` session variable is used in a row permission object to isolate data among tenants. The variable is created in `FHIRSchemaGenerator.addVariable`, and its definition should not be modified. 

> The variable privileges (grants) are updated one time, and applied to `FHIRUSER` user. The `FHIRUSER` gets only read access to this variable. To set the value of the variable, `FHIRUSER` must call `FHIR_ADMIN.SET_TENANT` as described in the next section.

Reviewer: ```suggestion
The variable privileges (grants) are updated one time, and applied to `FHIRSERVER` user. The `FHIRSERVER` gets only read access to this variable. To set the value of the variable, `FHIRSERVER` must call `FHIR_ADMIN.SET_TENANT` as described in the next section.
```

Author: fixed
